src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -44,59 +50,93 @@ object VariantDataset {
-                row.getSeq[Row](2).lazyMap { rg =>
-                  new RowGenotype(rg): Genotype
-                }))
-          } else
+                Iterable.empty[Annotation])))
+        else {
+          val rdd = sqlContext.readParquetSorted(parquetFile)
           rdd.map { row =>
             val v = row.getVariant(0)
             (v,
               (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
-                row.getGenotypeStream(v, 2, isDosage): Iterable[Genotype]))
+                row.getSeq[Any](2).lazyMap { g => if (gRequiresConversion) SparkAnnotationImpex.importAnnotation(g, genotypeSignature) else g }
+              ))
           }
+        }
+
+        OrderedRDD(rdd, partitioner)
       }
 
-      val partitioner: OrderedPartitioner[Locus, Variant] =
-        try {
-          val jv = hConf.readFile(dirname + "/partitioner.json.gz")(JsonMethods.parse(_))
-          jv.fromJSON[OrderedPartitioner[Locus, Variant]]
-        } catch {
-          case _: FileNotFoundException =>
-            fatal("missing partitioner.json.gz when loading VDS, create with HailContext.write_partitioning.")
+      new VariantSampleMatrix[Annotation](hc, metadata, localValue, orderedRDD)
+    } else {
+      val orderedRDD = if (skipVariants)
+        OrderedRDD.empty[Locus, Variant, (Annotation, Iterable[Genotype])](sc)
+      else {
+        val rdd = if (skipGenotypes)
+          sqlContext.readParquetSorted(parquetFile, Some(Array("variant", "annotations")))
+            .map(row => (row.getVariant(0),
+              (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
+                Iterable.empty[Genotype])))
+        else {
+          val rdd = sqlContext.readParquetSorted(parquetFile)
+          if (parquetGenotypes)
+            rdd.map { row =>
+              val v = row.getVariant(0)
+              (v,
+                (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
+                  row.getSeq[Row](2).lazyMap { rg =>
+                    new RowGenotype(rg): Genotype
+                  }))
+            } else
+            rdd.map { row =>
+              val v = row.getVariant(0)
+              (v,
+                (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
+                  row.getGenotypeStream(v, 2, isDosage): Iterable[Genotype]))
+            }
         }
 
-      OrderedRDD(rdd, partitioner)
-    }
+        OrderedRDD(rdd, partitioner)
+      }
 
-    new VariantSampleMatrix[Genotype](hc,
-      if (skipGenotypes) metadata.copy(sampleIds = IndexedSeq.empty[String],
-        sampleAnnotations = IndexedSeq.empty[Annotation])
-      else metadata,
-      orderedRDD)
+      new VariantSampleMatrix[Genotype](hc, metadata, localValue, orderedRDD)
+    }
```

Comment:
Using Generators of size `10000`, I see no real difference:
```
size,type,only-sites,only-samples,mean,stddev
10000,generic ,N,N,7.1919088E8,4.8565437942708987E8
10000,generic ,Y,N,3.732453167E8,2.4752663470772764E8
10000,generic ,Y,Y,1.73290874E7,7103529.212736992
10000,generic ,N,Y,1.72581589E7,6705925.585800972
10000,vds     ,N,N,4.47685612E8,2.449563475331303E8
10000,vds     ,Y,N,3.123765853E8,1.4380390083184013E8
10000,vds     ,Y,Y,1.39913781E7,6026181.388363289
10000,vds     ,N,Y,1.37619788E7,5674273.974913633
10000,parquet ,N,N,3.227085415E8,1.355673429218226E8
10000,parquet ,Y,N,3.264297298E8,1.6973637966874427E8
10000,parquet ,Y,Y,1.16073288E7,3395572.1130666863
10000,parquet ,N,Y,1.01950797E7,2012614.0968836548

size,type,only-sites,only-samples,mean,stddev
10000,generic ,N,N,6.184223947E8,4.356005784948978E8
10000,generic ,Y,N,3.64943412E8,2.6714179991928625E8
10000,generic ,Y,Y,1.65759601E7,7471753.613193231
10000,generic ,N,Y,1.40700359E7,4967014.820041117
10000,vds     ,N,N,4.289718703E8,1.8140829837982917E8
10000,vds     ,Y,N,6.893006907E8,4.656587041191373E8
10000,vds     ,Y,Y,1.4039354E7,7422751.767325215
10000,vds     ,N,Y,1.38024957E7,5790176.024989983
10000,parquet ,N,N,3.199654616E8,1.4082570434827748E8
10000,parquet ,Y,N,2.959852927E8,1.484514530976039E8
10000,parquet ,Y,Y,1.16814165E7,3411132.2841117214
10000,parquet ,N,Y,1.12049913E7,3095423.0483088107
```

I used [this patch](https://gist.github.com/danking/e1ff5d2299f6d4b954f910e8a116e28f) to test my branch and your branch.